### ISSUE TITLE
sinon.match.each ?

### ISSUE THREAD

**Issue by cyrilchapon** [sinonjs/sinon#1624]
It would be very great to have a matcher that exactly act like `sinon.match.has` but on every enumerated value in an array / object

Use case :
```
          ...
          const res = httpMocks.createResponse();
          res.json = sinon.spy(res.json);

          return stuffController.listStuff(req, res, next)
          .then(function() {
            assert.calledWith(res.json, sinon.match.array);
            assert.calledWith(res.json, sinon.match.has('length', expectedStuff.length));
            assert.calledWith(res.json, sinon.match(function(result) {
              return _.every(result, function(stuff) {
                return stuff instanceof Stuff;
              });
            }, 'Every return object is an stuff'));
          });
```

That could be turned to
```
          ...
          const res = httpMocks.createResponse();
          res.json = sinon.spy(res.json);

          return stuffController.listStuff(req, res, next)
          .then(function() {
            assert.calledWith(res.json, sinon.match.array);
            assert.calledWith(res.json, sinon.match.has('length', expectedStuff.length));
            assert.calledWith(res.json, sinon.match.each(result, sinon.match.instanceOf(Stuff)));
          });
```

**Comment by cyrilchapon** [comment]
It could be named `sinon.match.each`, or lodash-like `sinon.match.every`

**Comment by cyrilchapon** [comment]
In addition, one could find useful to check that at least one enumerated value matches.
We could have `sinon.match.some` for this

**Comment by mroderick** [comment]
I think that would be a great addition to matchers.

My preference leans more towards using names that are more like [`Array.prototype.every`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every) and [`Array.prototype.some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some), than using lodash conventions. Keeping in mind just how inconsistent Sinon's API currently is, I don't think that preference matters that much.

**Comment by mroderick** [comment]
This has been implemented by #1661 and shipped as `sinon@4.2.0`
